### PR TITLE
NAS-115098 / 22.12 / Add ability to specify render_context for etc files

### PR DESCRIPTION
**Pull request by anodos325**
It's often the case that several members of the same etc files
group will perform identical queries. Add ability to specify
a general configure context (methods whose results should be
avaliable to all members of the etc_files group) to avoid
duplicate queries. This is helpful for cases where method
being called is expensive.

### CODE REVIEW

**Comment by anodos325** [comment]
Had thought about this before lunch and thought this sort of capability might be useful to have. Although not strictly necessary for NFS case (these particular queries are pretty fast), there are other cases where it may be beneficial. NFS is here is example of idea.

File: src/middlewared/middlewared/plugins/etc.py [hunk 1]
```diff
@@ -335,12 +339,25 @@ def __init__(self, *args, **kwargs):
             'py': PyRenderer(self),
         }
 
+    async def gather_context(self, methods):
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/etc.py:342]: Let's call it consistently: it's either `context` or `ctx`, not both

File: src/middlewared/middlewared/plugins/etc.py [hunk 2]
```diff
@@ -335,12 +339,25 @@ def __init__(self, *args, **kwargs):
...
+
+        return rv
+
     async def generate(self, name, checkpoint=None):
         group = self.GROUPS.get(name)
         if group is None:
             raise ValueError('{0} group not found'.format(name))
 
         async with self.LOCKS[name]:
+            if group[0]['type'] == 'ctx':
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/etc.py:355]: Looks a little bit hacky. Can we just change definition style for groups that require a context? Something like
```
        'nfsd': {
            'ctx': {'methods': ['sharing.nfs.query', 'nfs.config']}},
            'entries': [
                {'type': 'mako', 'path': 'default/nfs-common'},
                {'type': 'mako', 'path': 'default/nfs-kernel-server'},
            ],
```

**Review by themylogin — CHANGES_REQUESTED** [review]
Great idea!

File: src/middlewared/middlewared/plugins/etc.py (hunk 2 above)
Comment (anodos325) [reply src/middlewared/middlewared/plugins/etc.py:355]: Good idea.

**Review by themylogin — APPROVED** [review]
Super, thank you!

**Review by yocalebo — APPROVED** [review]

**Pull request merged**